### PR TITLE
Issue #361 - Add user-friendly error message mapping for qube-config.json loader

### PR DESCRIPTION
**Pull request by Giraffe-Technology**
Satisfies Issue #361.

### CODE REVIEW

File: csvcubed/csvcubed/cli/error_mapping.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        return msg
+
+    except KeyError as err:
+        print(err)
+        return str(error.message)
+
+
+
+    #
+    # if isinstance(error, CsvColumnUriTemplateMissingError):
```
Comment (robons) [inline csvcubed/csvcubed/cli/error_mapping.py:103]: Do we need this commented out stuff?

File: csvcubed/csvcubed/cli/error_mapping.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
+from typing import TypeVar
+from os import linesep
+
+from csvcubed.models.validationerror import *
+from csvcubed.models.cube.validationerrors import *
+from csvcubed.models.cube.qb.validationerrors import *
+from csvcubed.models.cube.qb.components.validationerrors import *
+
+# TValidationError = TypeVar("TValidationError", bound=ValidationError, covariant=True)
```
Comment (robons) [inline csvcubed/csvcubed/cli/error_mapping.py:9]: Do we need this type anymore?

File: csvcubed/csvcubed/cli/error_mapping.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        "A validation error occurred when validating the cube: '{error[message]}'."
+}
+
+
+def friendly_error_mapping(error: ValidationError) -> str:
+    """
+    Given an Exception / Error it returns an error message that is more user-friendly
+    """
+    try:
+        en = error.__class__.__name__
```
Comment (robons) [inline csvcubed/csvcubed/cli/error_mapping.py:67]: could you give these a bit more of a readable name, like `error_name` and `error_dict`?

File: csvcubed/csvcubed/cli/error_mapping.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        msg = MSG_TEMPLATE[error.__class__]
+        ed = error.as_dict()
+
+        if isinstance(error, SpecificValidationError):
```
Comment (robons) [inline csvcubed/csvcubed/cli/error_mapping.py:71]: Nothing being done with the value in this if statement? Do we need it?

File: csvcubed/csvcubed/cli/error_mapping.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            error.get_error_url()
+
+        if isinstance(error, ConflictingUriSafeValuesError):
```
Comment (robons) [inline csvcubed/csvcubed/cli/error_mapping.py:74]: I'm not 100% sure I like the special-casing here. 

Perhaps you could change `MSG_TEMPLATE` so that it's a `Dict[Type[ValidationError], Callable[[ValidationError], str]` - i.e. a map of the validation error type to a lambda which takes the error and returns a string (the error message). Then we could do all of the necessary mapping and have it controlled from one location.

If you need let's have a chat about this.

File: csvcubed/csvcubed/cli/error_mapping.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        if isinstance(error, (MaxNumComponentsExceededError,
+                              MinNumComponentsNotSatisfiedError,
+                              WrongNumberComponentsError,
+                              NeitherDefinedError,
+                              IncompatibleComponentsError)
+                      ):
+            if error.additional_explanation:
+                msg += f"\nFurther details are: {error.additional_explanation}"
+
+        if isinstance(error, SpecificValidationError):
```
Comment (robons) [inline csvcubed/csvcubed/cli/error_mapping.py:90]: Could you check that it's `HasErrorUrl` here instead of `SpecificValidationError`?

Comment (robons) [reply csvcubed/csvcubed/cli/error_mapping.py:90]: Also, can you make sure to use `os.linesep` so we're more confident in our platform-independence.

File: csvcubed/csvcubed/cli/error_mapping.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+            msg += f"\nRefer to {error.get_error_url()} for guidance on correcting this problem."
+
+
+        return msg
+
+    except KeyError as err:
+        print(err)
```
Comment (robons) [inline csvcubed/csvcubed/cli/error_mapping.py:97]: I don't see the value of this catch block here - if we left it to bubble up to the top then the exception would be correctly logged. Can we leave it like that?

File: csvcubed/tests/unit/utils/test_pandas.py [hunk 8]
```diff
@@ -14,7 +14,7 @@ def test_na_filter_works():
     running the 'read_csv' command and checking that some of the default na-values
     don't get transformed to NaN values.
     """
-    df = read_csv(csv_path)
+    df, _data_errors = read_csv(csv_path)
```
Comment (robons) [inline csvcubed/tests/unit/utils/test_pandas.py:17]: Do we have a unit test to ensure that we're returning an error when there are duplicate column titles?

File: csvcubed/tests/behaviour/steps/qbwriter.py [hunk 9]
```diff
@@ -145,7 +145,7 @@ def step_impl(context, cube_name: str, csvw_file_path: str):
     ]
 
     csv_path, _ = get_first_table_schema(csvw_path)
-    code_list_data = read_csv(csv = csvw_path.parent / csv_path)
+    code_list_data, _data_errors = read_csv(csv=csvw_path.parent / csv_path)
```
Comment (robons) [inline csvcubed/tests/behaviour/steps/qbwriter.py:148]: Again, we should be testing something about these errors. Ideally they should be stored in the context and validated in the `the CSVqb should pass all validations` step.

File: csvcubed/tests/test-cases/cli/build/cube_data_config_ok.csv [hunk 10]
```diff
@@ -0,0 +1,3 @@
+Dim-0,Dim-1,Dim-2,Attr-1,Amount,Measure,Units
```
Comment (robons) [inline csvcubed/tests/test-cases/cli/build/cube_data_config_ok.csv:1]: Why have all of these test-case files been added in this PR? Are they being used anywhere?

File: csvcubed/csvcubed/readers/cubeconfig/v1_0/configdeserialiser.py [hunk 11]
```diff
@@ -200,7 +211,7 @@ def _configure_remaining_columns_by_convention(
 
             ordered_columns.append(qb_column)
 
-    cube.columns = ordered_columns
+    cube.columns = ordered_columns + list(configured_columns.values())
```
Comment (robons) [inline csvcubed/csvcubed/readers/cubeconfig/v1_0/configdeserialiser.py:214]: didn't we make this change in a different PR that's already been merged into main?

File: csvcubed/csvcubed/readers/cubeconfig/v1_0/configdeserialiser.py [hunk 12]
```diff
@@ -57,28 +61,34 @@
...
-            schema_validation_errors = validate_dict_against_schema(
-                value=config, schema=schema
-            )
+            try:
+                schema = load_resource(schema_path)
+                schema_validation_errors = validate_dict_against_schema(
+                    value=config, schema=schema
+                )
+            except JSONDecodeError as err:
+                log.warning(
```
Comment (robons) [inline csvcubed/csvcubed/readers/cubeconfig/v1_0/configdeserialiser.py:88]: This bit looks good to me, thanks.

File: csvcubed/csvcubed/readers/cubeconfig/v1_0/configdeserialiser.py [hunk 13]
```diff
@@ -5,18 +5,22 @@
...
 from jsonschema.exceptions import ValidationError as JsonSchemaValidationError
 from csvcubedmodels.rdf.namespaces import GOV
+from pandas import DataFrame
 
 from csvcubed.models.cube import *
+from csvcubed.models.validationerror import ValidationError
 from .mapcolumntocomponent import map_column_to_qb_component
 from csvcubed.utils.dict import get_with_func_or_none
 from csvcubed.utils.iterables import first
+from csvcubed.utils.pandas import read_csv
```
Comment (robons) [inline csvcubed/csvcubed/readers/cubeconfig/v1_0/configdeserialiser.py:23]: Again, wasn't this stuff done in a previous PR and merged into main?

File: csvcubed/csvcubed/readers/cubeconfig/v1_0/columnschema.py [hunk 14]
```diff
@@ -334,7 +334,7 @@ def map_to_qb_observation(self) -> QbObservationValue:
             if isinstance(self.measure, str):
                 measure = ExistingQbMeasure(self.measure)
 
-            elif isinstance(self.measure, NewMeasures):
+            elif isinstance(self.measure, Measure):
```
Comment (robons) [inline csvcubed/csvcubed/readers/cubeconfig/v1_0/columnschema.py:337]: Good catch. We probably need another ticket to go back and ensure we've got enough test coverage in this area.

**Review by robons — CHANGES_REQUESTED** [review]
Again, there are a number of lines of commented out code which I suspect we don't want to merge into main.

More concerningly, there don't seem to be any new tests along with this code. Can you make sure to add some?

Going forwards, can you make sure you're checking your PRs for things like these, please?

File: csvcubed/csvcubed/cli/build.py [hunk 15]
```diff
@@ -68,14 +68,17 @@ def build(
...
 
             with open(validation_errors_file_out, "w+") as f:
-                json.dump(all_errors, f, indent=4)
+                json.dump(all_errors, f, indent=4, default=serialize_sets)
+            # with open(output_directory / validation_errors_file_out, "w+") as f:
+            #     json.dump(all_errors, f, indent=4, default=serialize_sets)
 
         if fail_when_validation_error_occurs and len(validation_errors) > 0:
             exit(1)
+    else:
```
[the remark this answers is not in the document]
Comment (Giraffe-Technology) [reply csvcubed/csvcubed/cli/build.py:77]: When a cube is built but validation_errors have been returned, its not possible to write the cube using QbWriter as exceptions are raised. This change was to avoid this happening.  It would be better to wrap this in a try-except block or to update QbWriter to better handle known exception cases.  
e.g.
self = QbWriter(cube=Cube(metadata=CatalogMetadata(title='Tests/test-cases/config/schema-cube-data-config-ok', identifier='sc...be=Cube(metadata=CatalogMetadata(title='Tests/test-cases/config/schema-cube-data-config-ok', identifier='schema-id'))))
t = <class 'csvcubed.models.cube.qb.components.measuresdimension.QbMultiMeasureDimension'>

    def _get_single_column_of_type(
        self, t: Type[QbColumnarDsdType]
    ) -> QbColumn[QbColumnarDsdType]:
        cols = get_columns_of_dsd_type(self.cube, t)
        if len(cols) != 1:
>           raise ValueError(
                f"Found {len(cols)} columns with component type {t} in cube. Expected 1."
            )
E           ValueError: Found 0 columns with component type <class 'csvcubed.models.cube.qb.components.measuresdimension.QbMultiMeasureDimension'> in cube. Expected 1.

../../csvcubed/writers/qbwriter.py:943: ValueError

File: csvcubed/tests/test-cases/cli/build/cube_data_config_ok.csv (hunk 10 above)
Comment (Giraffe-Technology) [reply csvcubed/tests/test-cases/cli/build/cube_data_config_ok.csv:1]: These files were being used by the test_build_friendly_error_msgs.py unit tests which I failed to commit.
I've now cleaned out a few files and updated the file names to match the exception being tested for and moved them
to tests/test-cases/readers/cube-config/v1.0/error_mappings folder.

File: csvcubed/csvcubed/readers/cubeconfig/v1_0/columnschema.py (hunk 14 above)
Comment (Giraffe-Technology) [reply csvcubed/csvcubed/readers/cubeconfig/v1_0/columnschema.py:337]: Issue raised https://app.zenhub.com/workspaces/features-squad-61b72275ac896c0010a1b00b/issues/gss-cogs/csvcubed/427.

File: csvcubed/csvcubed/readers/cubeconfig/v1_0/configdeserialiser.py (hunk 11 above)
Comment (Giraffe-Technology) [reply csvcubed/csvcubed/readers/cubeconfig/v1_0/configdeserialiser.py:214]: No, its not in main,  we did work on this together for this task.

**Review by robons — APPROVED** [review]

**Review by canwaf — APPROVED** [review]
The mystery machine solved the haunted test!
